src/app-router.js

```diff
@@ -40,53 +34,37 @@ export class AppRouter extends Router {
     });
   }
 
-  dequeueInstruction() {
-    if (this.isNavigating) {
-      return;
-    }
-
-    var instruction = this.queue.shift();
-    this.queue = [];
-
-    if (!instruction) {
-      return;
-    }
-
-    this.isNavigating = true;
-    this.events.publish('router:navigation:processing', instruction);
-
-    var context = this.createNavigationContext(instruction);
-    var pipeline = this.pipelineProvider.createPipeline(context);
+  dequeueInstruction(instructionCount = 0) {
+    return Promise.resolve().then(() => {
+      if (this.isNavigating && !instructionCount) {
```

Comment:
"inner" instructions, such as redirects and instructions added to the queue during other navigations are now considered part of the "outer" navigation, instead of completing the current instruction before moving on to the next.